system-contracts/contracts/erc4337/EntryPointV01.sol

```diff
@@ -0,0 +1,168 @@
+        _payForTransaction(txHash, suggestedTxHash, tx);
+
+        // TODO: here and below, we probably should not revert after the transaction payment;
+        // instead we should go to the next transaction.
+        _executeTransaction(txHash, suggestedTxHash, tx);
+
+        // refund?
+    }
+
+    function _validateTransaction(bytes32 _txHash, bytes32 _suggestedSignedHash, Transaction memory _tx) private {
+        address from = address(uint160(_tx.from));
+        // check account type via ContractDeployer
+        IContractDeployer.AccountAbstractionVersion version = ContractDeployer(address(DEPLOYER_SYSTEM_CONTRACT))
+            .extendedAccountVersion(from);
+        require(version == IContractDeployer.AccountAbstractionVersion.Version1, "Unsupported account version");
+
+        // check that nonce is available yet
+        NONCE_HOLDER_SYSTEM_CONTRACT.validateNonceUsage(from, _tx.nonce, false);
+
+        // validate transaction
+        bytes memory returnData = this._performMimicCall(
+            uint32(gasleft()), // Should be value from the transaction?
+            BOOTLOADER_FORMAL_ADDRESS,
+            from,
+            abi.encodeCall(IAccount(from).validateTransaction, (_txHash, _suggestedSignedHash, _tx))
+        );
+        bytes4 magic = abi.decode(returnData, (bytes4));
+        // We have to revert, since user didn't pay for the transaction just yet.
+        require(magic == IAccount.validateTransaction.selector, "Verification failed");
+
+        // check that nonce is not available anymore
+        NONCE_HOLDER_SYSTEM_CONTRACT.validateNonceUsage(from, _tx.nonce, true);
+    }
+
+    function _payForTransaction(bytes32 _txHash, bytes32 _suggestedSignedHash, Transaction memory _tx) private {
+        address from = address(uint160(_tx.from));
+        if (_tx.paymaster == 0) {
+            // No paymaster, pay directly
+            uint256 bootloaderBalanceBefore = address(BOOTLOADER_FORMAL_ADDRESS).balance;
+            this._performMimicCall(
+                uint32(gasleft()), // Should be value from the transaction?
+                BOOTLOADER_FORMAL_ADDRESS,
+                from,
+                abi.encodeCall(IAccount(from).payForTransaction, (_txHash, _suggestedSignedHash, _tx))
+            );
+            uint256 bootloaderBalanceAfter = address(BOOTLOADER_FORMAL_ADDRESS).balance;
+            require(bootloaderBalanceAfter > bootloaderBalanceBefore, "Transaction payment failed");
+            require(
+                bootloaderBalanceAfter - bootloaderBalanceBefore >= _tx.gasLimit * _tx.maxFeePerGas,
+                "Transaction payment amount mismatch"
+            );
+            // TODO: should we send back excessive funds like bootloader does?
+        } else {
+            // Pay through the paymaster
+            revert("Not implemented yet");
+        }
+    }
+
+    function _executeTransaction(bytes32 _txHash, bytes32 _suggestedSignedHash, Transaction memory _tx) private {
```

Comment:
this function does not handle user tx failing, though I guess it is the case in the initial iteration only